src/python/pants/backend/python/goals/setup_py.py

```diff
@@ -443,33 +443,48 @@ async def run_setup_pys(
             f'{", ".join(so.address.spec for so in explicit_nonexported_targets)}'
         )
 
+    transitive_targets_per_exported_target = await MultiGet(
+        Get(TransitiveTargets, TransitiveTargetsRequest([et.target.address]))
+        for et in exported_targets
+    )
+
     if setup_py_subsystem.transitive:
-        # Expand out to all owners of the entire dep closure.
-        transitive_targets = await Get(
-            TransitiveTargets,
-            TransitiveTargetsRequest(et.target.address for et in exported_targets),
+        closure = FrozenOrderedSet(
+            itertools.chain.from_iterable(
+                tt.closure for tt in transitive_targets_per_exported_target
+            )
         )
         owners = await MultiGet(
             Get(ExportedTarget, OwnedDependency(tgt))
-            for tgt in transitive_targets.closure
+            for tgt in closure
             if is_ownable_target(tgt, union_membership)
         )
         exported_targets = list(FrozenOrderedSet(owners))
+        transitive_targets_per_exported_target = await MultiGet(
```

Comment:
Haven't you already computed this above?

Reply:
We might have more `exported_targets` than we had before, note that we're redefining `exported_targets`. The `zip()` wouldn't be accurate anymore.